### PR TITLE
Fix cronjob status reconciliation when job template labels change

### LINKED ISSUE #106496: CronJob loses track of Jobs after label update
### What happened?

Changing the label on a job template in a CronJob causes any Jobs held by the successful/failed history parameters to stop being tracked, so they never get cleaned up.

### What did you expect to happen?

Jobs with labels that don't match the (upgraded) CronJob, but still have an owner reference, should be subject to the history limit.

### How can we reproduce it (as minimally and precisely as possible)?

1. Deploy a CronJob with a job template containing label `foo: one`, and non-zero history limit
2. Allow the CronJob to run and generate some Jobs (default 3 successful, 1 failed)
3. Update the CronJob with a job template containing label `foo: two`
4. Allow the CronJob to run and generate some more Jobs
5. Notice there are more than 3 successful Jobs/1 failed Job!

### Anything else we need to know?

I believe this is the relevant code: https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/cronjob/cronjob_controllerv2.go#L235-L239

The CronJob controller finds Jobs that it should care about by matching on the labels in the job template, this obviously breaks if the labels in the job template change.

### Kubernetes version

1.21

### Cloud provider

Self-hosted

### OS version

_No response_

### Install tools

_No response_

### Container runtime (CRI) and and version (if applicable)

_No response_

### Related plugins (CNI, CSI, ...) and versions (if applicable)

_No response_

### PR DESCRIPTION
**Pull request by d-honeybadger**
#### What type of PR is this?
/kind bug
/kind regression

#### What this PR does / why we need it:
In CronJobController V2, the controller only considers jobs with labels that match current cronjob's job template labels. This leads to the controller [losing track of jobs owned by a given cronjob when those template labels change](https://github.com/kubernetes/kubernetes/issues/106496). For example, if job template labels were updated while the cronjob was in the `active` status, that cronjob will be stuck as `active`, since the controller will never find the completed job whose labels now don't match those of the job template.

#### Which issue(s) this PR fixes:
Fixes #106496

#### Special notes for your reviewer:
I see that in https://github.com/kubernetes/kubernetes/pull/93370 filtering by label was added for performance. If anyone has ideas on how to keep that performance improvement while maintaining correct behavior, I'd be happy to close this and work on another solution.

#### Does this PR introduce a user-facing change?
```release-note
Fixes regression in 1.21 default configurations where a bug in CronJob Controller V2 where it would lose track of jobs upon job template labels change.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by d-honeybadger** [comment]
\assign @soltysh

**Comment by alculquicondor** [comment]
/ok-to-test

This definitely requires a release note.

**Comment by alculquicondor** [comment]
For comparison, in the job controller we list all the pods for a similar reason: https://github.com/kubernetes/kubernetes/blob/3866cb91f22da6eb49dab10dd5c33385690b57b4/pkg/controller/job/job_controller.go#L558-L560

So I don't see why we shouldn't do the same for the cronjob controller.

I'll have to defer to @soltysh as I don't have the historical background.

**Comment by d-honeybadger** [comment]
> This definitely requires a release note.

Ah yes, added.
Thank you for taking a look!

**Comment by alculquicondor** [comment]
You should probably add a test too :)
But maybe wait for Maciej to confirm if the change is reasonable.

**Comment by janetkuo** [comment]
> For comparison, in the job controller we list all the pods for a similar reason:
> 
> https://github.com/kubernetes/kubernetes/blob/3866cb91f22da6eb49dab10dd5c33385690b57b4/pkg/controller/job/job_controller.go#L558-L560
> 
> So I don't see why we shouldn't do the same for the cronjob controller.
> 
> I'll have to defer to @soltysh as I don't have the historical background.

The reason that we didn't do this for the original CronJob controller is that it doesn't use watches, see the [ControllerRef design doc](https://github.com/kubernetes/design-proposals-archive/blob/main/api-machinery/controller-ref.md#controller-specific-behavior) for more details. Given that CronJob v2 now uses watches, it makes sense to make the changes accordingly.

**Comment by alculquicondor** [comment]
Makes sense.

@d-honeybadger could you add a unit test for the change?

**Review by soltysh — APPROVED** [review]
Good catch, I'll dig through perf data and see how good/bad this is, alternatively we could add a separate indexer for jobs, if we needed, but that can be done separately from this PR.
But I definitely would like to see unit test added like @alculquicondor mentioned
/approve
/triage accepted
/priority important-longterm
/assign @alculquicondor 
he has the final lgtm on it

**Comment by d-honeybadger** [comment]
@alculquicondor :+1:  Added a test and a comment for good measure.

File: pkg/controller/cronjob/cronjob_controllerv2_test.go [hunk 1]
```diff
@@ -1067,6 +1067,30 @@ func TestControllerV2GetJobsToBeReconciled(t *testing.T) {
...
+					ObjectMeta: metav1.ObjectMeta{Labels: map[string]string{"key": "value"}},
+				}},
+			},
+			jobs: []runtime.Object{
+				&batchv1.Job{ObjectMeta: metav1.ObjectMeta{
+					Namespace:       "foo-ns",
+					Name:            "foo-nonmatching-label",
+					Labels:          map[string]string{"key": "different-value"},
+					OwnerReferences: []metav1.OwnerReference{{Name: "fooer", Controller: &trueRef}}},
+				},
```
Comment (alculquicondor) [inline pkg/controller/cronjob/cronjob_controllerv2_test.go:1084]: Add one that has a different owner reference.

**Review by alculquicondor — COMMENTED** [review]
/lgtm

Thanks!

**Comment by simonjpartridge** [comment]
Just curious if this bug will get backported to any older versions of k8s? Or just released in the 1.24 release?

**Pull request merged**